### DISCUSSION TITLE
RegexpSingleline/RegexpSinglelineJava  - No column in error message

### DISCUSSION THREAD

**Discussion by VirtualTim** [checkstyle/checkstyle#12818] [Q&A] [answered]
I'm started using RegexpSinglelineJava to catch invalid log messages*, but violations of this check report only a line number, not a column number. I could hoping that capture groups would have some relation with column numbers, but that doesn't seem to be the case.
An I missing something, or is this a bug/enhancement?

* The log library I'm using uses java.text.MessageFormat, so I'm checking for things like `{}` or `'{0}'`.

**Comment by romani** [comment, marked as answer]
We match line by line, we never thought about matching and printing columns of all matching positions of groups. If you see how can we improve logging of violation , please let us know.
If Java implmnetation of regexp able to get columns for groups, we just need to figureout how to make fancy formatting of message to print columns is user required. Can you write some simple java code that does this, we print messages by `String.format`.
We need some flexible parameters.... uh ...... It might be  better to write custom applicaiton to do this as you need after checkstyle validation, based on checkstyle report.

**Reply by VirtualTim** [reply]
> If Java implmnetation of regexp able to get columns for groups, we just need to figureout how to make fancy formatting of message to print columns is user required.

Java has `matcher(string).start(groupNumber)`. Documentation for `start`:
> Returns the start index of the subsequence captured by the given group during the previous match operation.
Capturing groups are indexed from left to right, starting at one. Group zero denotes the entire pattern, so the expression m.start(0) is equivalent to m.start().

So code could look something like this:
```Java
int column = matcher(string).start();
if (column == -1) {
  //Use current String.format(...)
} else {
  //A group matched, so use that.
  //If a user does not want the first group found to be the column number, they should use non-capturing groups
  String.format("%s at %s:%d:%d", message, file, row column );
}
```

**Reply by VirtualTim** [reply]
So I don't think this would be hard to achieve.

**Reply by romani** [reply]
https://checkstyle.org/config_regexp.html#RegexpSinglelineJava_Violation_Messages
https://github.com/checkstyle/checkstyle/blob/13481f2c410e4944ecf5ab93ec49948a523a0c82/src/main/resources/com/puppycrawl/tools/checkstyle/checks/regexp/messages.properties#L4

as you can see message format is stict and simple , we need to figure out how to make it to print what you need on unlimited amount of groups.

**Reply by VirtualTim** [reply]
Interesting, other checks don't specify a column in their messages.properties, so I wonder where that comes from?

**Reply by romani** [reply]
Columns come from Checks that are based on ASTree , parsed content. 

Regexp Check is like grep, no understanding that it is java

**Comment by nrmancuso** [comment]
@VirtualTim please share your regular expression, we might be able to steer you towards another check that could help.

**Reply by VirtualTim** [reply]
My Regex is: `"\b(log|LOG).*(?:(\{\})|(?:([^'])'([^'])))"`, and here's my reasoning:
> Regex to catch MessageFormat quirks:
       1. Replacement syntax uses {\d}, not {}. This is different from the more common slf4j.
       2. Single quites are treated as escape characters.

**Reply by romani** [reply]
it will be much easier for you to write custom Check
that applies xpath to find String literal, and apply regexp to it
maybe some modification of https://checkstyle-addons.thomasjensen.com/v7.0.1/checks/regexp.html#RegexpOnString

keep in mind that loging messages a long sometime, so you need to be aware of concatenation by `+` in code, to search for in whole string value, not a first part of it.
